test/e2e/runner/infra.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"context"
+
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+)
+
+// TestnetInfra provides an API for interacting with the infrastructure for an
+// entire testnet.
+type TestnetInfra interface {
+	// GenerateConfig generates any necessary configuration for the
+	// infrastructure provider during testnet setup.
+	GenerateConfig() error
+
+	// Stop attempts to stop the entire testnet.
+	Stop(ctx context.Context) error
+
+	// Pause attempts to pause the entire testnet.
+	Pause(ctx context.Context) error
+
+	// Unpause attempts to resume a paused testnet.
+	Unpause(ctx context.Context) error
+
+	// ShowLogs will print all logs for the whole testnet to stdout.
+	ShowLogs(ctx context.Context) error
+
+	// TailLogs tails the logs for all nodes in the testnet.
+	TailLogs(ctx context.Context) error
+
+	// Cleanup stops and destroys all running infrastructure and deletes any
+	// generated files.
+	Cleanup(ctx context.Context) error
+}
+
+// NodeInfra provides an API for interacting with specific nodes'
+// infrastructure.
+type NodeInfra interface {
+	// ProvisionNode attempts to provision infrastructure for the given node
+	// and starts it.
+	ProvisionNode(ctx context.Context, node *e2e.Node) error
+
+	// DisconnectNode attempts to ensure that the given node is disconnected
+	// from the network.
+	DisconnectNode(ctx context.Context, node *e2e.Node) error
+
+	// ConnectNode attempts to ensure that the given node is connected to the
+	// network.
+	ConnectNode(ctx context.Context, node *e2e.Node) error
+
+	// KillNode attempts to ensure that the given node's process is killed
+	// immediately using SIGKILL.
+	KillNode(ctx context.Context, node *e2e.Node) error
+
+	// StartNode attempts to start a node's process. Assumes that the node's
+	// infrastructure has previously been provisioned using ProvisionNode.
+	StartNode(ctx context.Context, node *e2e.Node) error
```

Comment:
Could we require the implementation to report an error if the node wasn't provisioned? Or alternatively, could we require that StartNode call Provision automatically if it hasn't been? (Or maybe even just fold the two together?)

Reply:
I've now renamed this to `StartNodeProcess` - it's a specific bit of functionality required by the perturbation logic where we shouldn't be provisioning any new infrastructure.